Added toolchain project for cross compiling.

oopsmonk opened the pull request:
Fixing #16 
Added default cross toolchains include aarch64(armv8), armv7, x86_64, i686, and a python script to generate bazel toolchain files (CROSSTOOL, BUILDs).

Aarch64 Testing:
`bazel build --crosstool_top=@iota_toolchains//tools/aarch64--glibc--stable-2018.02-2:toolchain --cpu=aarch64 --host_crosstool_top=@bazel_tools//tools/cpp:toolchain --verbose_failures //cclient:api`

Commit:
Added toolchain project for cross compiling.

Commit:
Added toolchain project for cross compiling.

Commit:
Add i686 test to CI.

Commit:
Add i686 test to CI.

Commit:
Update toolchain gcc version to 8.1

Commit:
Update toolchain gcc version to 8.1

Commit:
Changed gcc version to 8.1 for CI.

Commit:
Changed gcc version to 8.1 for CI.

jserv commented:
By the way, can you provide instructions to re-generate toolchain from scratch? Something like [crosstool-NG](https://crosstool-ng.github.io/) might work out of box.

th0br0 commented on WORKSPACE:
```diff
@@ -6,6 +6,18 @@ git_repository(
 )
 
+git_repository(
+    name = "iota_toolchains",
```
remove

th0br0 commented:
@jserv that is available here https://github.com/bootlin/toolchains-builder

th0br0 commented:
```
/root/.cache/bazel/_bazel_root/ec321eb2cc2d0f8f91b676b6d4c66c29/external/bootlin_i686_toolchain/bin/../lib/gcc/i686-buildroot-linux-gnu/8.1.0/../../../../i686-buildroot-linux-gnu/bin/ld: bazel-out/i686-opt/bin/common/trinary/libtryte_long.a(tryte_long.o): in function `long_to_trytes':
--
  | /proc/self/cwd/common/trinary/tryte_long.c:57: undefined reference to `pow'
```

You're missing parameters like the below from your `CROSSTOOL` files:
```
  cxx_flag: "-std=c++11"
  linker_flag: "-lpthread"
  linker_flag: "-L/usr/local/llvm_scratch/libcxx_msan/lib"
  linker_flag: "-lc++"
  linker_flag: "-lc++abi"
  linker_flag: "-Wl,-rpath,/usr/local/llvm_scratch/libcxx_msan/lib"
  linker_flag: "-lm"
  linker_flag: "-fuse-ld=gold"
  linker_flag: "-B/usr/bin"
  linker_flag: "-B/usr/bin"
  linker_flag: "-Wl,--build-id=md5"
  linker_flag: "-Wl,--hash-style=gnu"
  cxx_builtin_include_directory: "/usr/local/llvm_scratch/libcxx_msan/include/c++/v1"
  cxx_builtin_include_directory: "/usr/local/llvm_scratch/libcxx_msan/include/"
  cxx_builtin_include_directory: "/usr/include/clang/4.0.0/include"
  cxx_builtin_include_directory: "/usr/lib/clang/4.0.0/include"
  cxx_builtin_include_directory: "/usr/lib/llvm-4.0/lib/clang/4.0.0/include"
  cxx_builtin_include_directory: "/usr/local/include"
  cxx_builtin_include_directory: "/usr/include/x86_64-linux-gnu"
  cxx_builtin_include_directory: "/usr/include"
```

Commit:
Fixed duplicate repo and linker errors.

duplicate git_repostory in WORKSPACE and linker errors in cross tools.

Commit:
Fixed duplicate repo and linker errors.

duplicate git_repostory in WORKSPACE and linker errors in cross tools.

oopsmonk commented:
The test function has an error in gcc 8.1, but with no problem in gcc 7.3. 
GCC-8  checks the comparison function with static assertions. 
See https://gcc.gnu.org/gcc-8/porting_to.html#comparison-object
```
cppclient/tests/api_json.cc:50:3:   required from here
bazel-out/i686-opt/bin/external/nlohmann/_virtual_includes/json/nlohmann/detail/conversions/to_json.hpp:312:45: error: static assertion failed: could not find to_json() method in T's namespace
         static_assert(sizeof(BasicJsonType) == 0,
~~~~~~~~~~~~~~~~~~~~~~^~~~
```

Commit:
Remove i686 toolchain test in CI temporally.

Some compile issues are needed to fix in gcc 8.1, then we can trun on the
toolchain test.

Commit:
Remove i686 toolchain test in CI temporally.

Some compile issues are needed to fix in gcc 8.1, then we can trun on the
toolchain test.

th0br0 committed:
common/trinary: fix trit_array_5 bug

th0br0 committed:
common/trinary: fix trit_array_5 bug

th0br0 committed:
cppclient/tests: fix googlemock & json issue

th0br0 committed:
cppclient/tests: fix googlemock & json issue

Commit:
Added i686 gcc8.1 toolchain build case to CI.

th0br0 committed:
cppclient/tests: fix api_json test again

th0br0 committed:
Merge remote-tracking branch 'samchen/crosstools' into crosstools

* samchen/crosstools:
  Added i686 gcc8.1 toolchain build case to CI.
  cppclient/tests: fix googlemock & json issue
  common/trinary: fix trit_array_5 bug
  Remove i686 toolchain test in CI temporally.
  Fixed duplicate repo and linker errors.
  Changed gcc version to 8.1 for CI.
  Update toolchain gcc version to 8.1
  Add i686 test to CI.
  Added toolchain project for cross compiling.

th0br0 committed:
buildkite: specify queue=dev
